akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala

```diff
@@ -1105,36 +1106,51 @@ class DDataShardCoordinator(
   val CoordinatorStateKey = LWWRegisterKey[State](s"${typeName}CoordinatorState")
   val initEmptyState = State.empty.withRememberEntities(settings.rememberEntities)
 
-  val AllShardsKey = GSetKey[String](s"shard-${typeName}-all")
-  val allKeys: Set[Key[ReplicatedData]] =
-    if (rememberEntities) Set(CoordinatorStateKey, AllShardsKey) else Set(CoordinatorStateKey)
-
-  var shards = Set.empty[String]
   var terminating = false
   var getShardHomeRequests: Set[(ActorRef, GetShardHome)] = Set.empty
 
-  if (rememberEntities)
-    replicator ! Subscribe(AllShardsKey, self)
+  val rememberEntitiesStore =
+    if (rememberEntities) {
+      rememberEntitiesStoreProvider.map(provider =>
+        context.actorOf(provider.coordinatorStoreProps(), "RememberEntitiesStore"))
+    } else None
 
   node.subscribe(self, ClusterEvent.InitialStateAsEvents, ClusterShuttingDown.getClass)
 
   // get state from ddata replicator, repeat until GetSuccess
   getCoordinatorState()
-  getAllShards()
+  if (settings.rememberEntities)
+    getAllRememberedShards()
 
-  override def receive: Receive = waitingForState(allKeys)
+  override def receive: Receive =
+    waitingForInitialState(
+      None,
+      if (rememberEntities) None
+      else Some(Set.empty))
 
   // This state will drop all other messages since they will be retried
-  def waitingForState(remainingKeys: Set[Key[ReplicatedData]]): Receive =
+  // Note remember entities query retries forever in store rather than here
+  def waitingForInitialState(initialState: Option[State], initialShards: Option[Set[ShardId]]): Receive =
     ({
+
+      case RememberEntitiesCoordinatorStore.RememberedShards(shardIds) =>
```

Comment:
I don't think waiting for `RememberedShards` is mandatory for the initialization process. Should be possible to do in the background. That can be explored in a separate PR, and maybe not important.

Reply:
Keeping current impl for now, created https://github.com/akka/akka/issues/28925 to track